docs/community/rust-coding-guidelines/rust-secure-coding-guidelines.md

```diff
@@ -0,0 +1,253 @@
+
+- Implement the Drop trait selectively, only when necessary for specific destructor logic. It's mainly used for managing external resources or memory in structures like Box or Rc, often involving unsafe code and security-critical operations.
+- In a Rust secure development, the implementation of the `std::ops::Drop` trait
+  must not panic.
+- Do not rely on `Drop` trait in security material treatment after the use, use [zeroize](https://docs.rs/zeroize/latest/zeroize/#) to explicit destroy security material, e.g. private keys.
+
+### Send and Sync Traits
+
+Be cautious with manual implementations of `Send` and `Sync` traits [[Rustbook: typesystem]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#send-and-sync-traits) [[Rustbook: send and sync]](https://doc.rust-lang.org/nomicon/send-and-sync.html).
+Both traits are _unsafe traits_, i.e., the Rust compiler does not verify in any way that they are implemented correctly. The danger is real: an incorrect implementation may lead to **undefined behavior**.
+
+In the majority of scenarios, manual implementation is unnecessary. In Rust, nearly all primitive types intrinsically implement Send and Sync traits, and for a significant proportion of compound types, the Rust compiler automatically derives these implementations.
+
+### Comparison Traits
+
+Ensure the implementation of standard comparison traits respects documented invariants [[comparison traits partialeq eq partialord ord]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#comparison-traits-partialeq-eq-partialord-ord).
+
+> The ANSSI resource extensively covers the matter, an example in [Appendix](#implementation-of-comparison-traits) is provided as a really basic demonstration of implementation.
+
+### Enums
+
+Prefer enums for state management to prevent invalid state representation.
+
+### Concurrency Safe Primitives
+
+Make use of Rust’s concurrency primitives like `Arc`, `Mutex`, and `RwLock` to manage shared state [[Rustbook: concurrency]](https://doc.rust-lang.org/book/ch16-00-concurrency.html).
+By utilizing these primitives, Rust programs can manage shared resources among multiple threads safely and efficiently, adhering to Rust's goals of enabling fearless concurrency and memory safety. This is critical in multithread programs because mistakes in concurrency can be very costly, both for the system's stability and security.
+
+### Data Structures with Deterministic Internal Order
+
+Certain data structures, like HashMap and HashSet, do not guarantee a deterministic order for the elements stored within them. This lack of order can lead to problems in operations that require processing elements in a consistent sequence across multiple executions. Below is a list of deterministic data structures available in Rust. Please note, this list may not be exhaustive:
```

Comment:
can we list some common scenarios in aptos when deterministic data structures are important?

Reply:
Added: In the Aptos blockchain, deterministic data structures help in achieving consensus, maintaining the integrity of the ledger, and ensuring that computations can be reliably reproduced across different nodes.